Core/Src/CAN/general_can.c

```diff
@@ -51,7 +51,7 @@ void CAN_general_receive_callback(FDCAN_HandleTypeDef* hfdcan, uint32_t RxFifo0I
         for (int i = 0; i < 4; ++i) {
             general_can_vars_e var_name = general_get_var_name_from_id_and_pos(id, i);
 
-            if ((int)var_name != -1) {
+            if (var_name < NUM_STATES_GENERAL) {
```

Comment:
Vc chegou a testar essa verificação? Quando tava testando a aquisição dos dados dos Inversores, via um comportamento estranho no código antigo. O var_name recebia 255 ao invés de -1 no retorno da Xmacro. Por conta disso, mesmo o ID e Pos não sendo listado, entrava no if. O jeito que fiz pra consertar isso foi assim:
                           if (var_name != (can_vars_inverter_e)255)
Aí não entrou mais no if quando o ID não era esperado

Reply:
A explicação do que aconteceu é mais interessante do que eu pensava. 

O motivo principal é essa frase aqui em baixo, [presente no manual do GCC](https://gcc.gnu.org/onlinedocs/gcc-8.2.0/gcc/Structures-unions-enumerations-and-bit-fields-implementation.html#Structures-unions-enumerations-and-bit-fields-implementation), o compilador que usamos: 

> Normally, the type is unsigned int if there are no negative values in the enumeration, otherwise int. If -fshort-enums is specified, then if there are negative values it is the first of signed char, short and int that can represent all the values, otherwise it is the first of unsigned char, unsigned short and unsigned int that can represent all the values.

Pelo que pesquisei essa flag `-fshort-enums` é ativada por default no gcc pra arm.

A `can_vars_inverter_e` é definida sem valor negativo, então ela é um uint8_t. A `general_get_var_name_from_id_and_pos` retorna `can_vars_inverter_e`, ou seja, um uint8_t. Porém, ela retorna -1 para caso de erro, que vira 255, pois novamente, ela retorna uint8_t. 

Quem implementou esse código colocou o cast para (int) pensando que ele ia considerar 255 como -1 (se o cast fosse pra uint8_t ia dá bom). Quem fez isso (e eu e Renan que aprovamos) provavelmente pensou que o enum seria int (uint32_t), pois isso é o default em vários compiladores.

O perigoso de C é que ele deixa a gente fazer coisa errada e nem avisa direito. Eu mexi nisso pra tirar o seguinte aviso:

> comparison is always true due to limited range of data type [-Wtype-limits]

Em que ele tá reclamando que o valor do if sempre será true, que era o problema que você passou. Mas ele não reclama nenhum outro erro que estava presente:

1. Retornamos um valor numérico ao invés da enum;
2. Esse valor numérico é negativo e a função retorna unsigned;
3. Comparamos um valor unsigned com um negativo;
4. O cast transforma um valor que era pra ser de 1 byte em um de 4 bytes e ferra com seu valor que já tá ferrado por causa de 2.

Eu fiz essa mudança aí pensando só no warning que ele me deu, e com isso resolvi apenas o erro 3, você fez a mesma coisa. Para resolver todos os erros vou criar mais um elemento do enum que equivale a -1 e fazer a comparação com todos os lados.